src/InlineRelationsTransformer.cpp

```diff
@@ -0,0 +1,965 @@
+// If unification is not successful, the returned vector is marked as invalid.
+// Assumes that the atoms are both of the same relation.
+NullableVector<std::pair<AstArgument*, AstArgument*>> unifyAtoms(AstAtom* first, AstAtom* second) {
+    std::vector<std::pair<AstArgument*, AstArgument*>> substitution;
+
+    std::vector<AstArgument*> firstArgs = first->getArguments();
+    std::vector<AstArgument*> secondArgs = second->getArguments();
+
+    // Create the initial unification equalities
+    for (int i = 0; i < firstArgs.size(); i++) {
+        substitution.push_back(std::make_pair(firstArgs[i], secondArgs[i]));
+    }
+
+    // Reduce the substitutions
+    bool success = reduceSubstitution(substitution);
+    if (success) {
+        return NullableVector<std::pair<AstArgument*, AstArgument*>>(substitution);
+    } else {
+        // Failed to unify the two atoms
+        return NullableVector<std::pair<AstArgument*, AstArgument*>>();
+    }
+}
+
+// Inlines the given atom based on a given clause.
+// Returns the vector of replacement literals and the necessary constraints.
+// If unification is unsuccessful, the vector of literals is marked as invalid.
+std::pair<NullableVector<AstLiteral*>, std::vector<AstConstraint*>> inlineBodyLiterals(
+        AstAtom* atom, AstClause* atomInlineClause) {
+    bool changed = false;
+    std::vector<AstLiteral*> addedLits;
+    std::vector<AstConstraint*> constraints;
+
+    // Start off by renaming the variables in the inlined clause to avoid conflicts when unifying multiple
+    // atom
+    // - particularly when an inlined relation appears twice in a clause.
+    static int inlineCount = 0;
+
+    // Make a temporary clone so we can rename variables without fear
+    AstClause* atomClause = atomInlineClause->clone();
+
+    struct VariableRenamer : public AstNodeMapper {
+        int varnum;
+        VariableRenamer(int varnum) : varnum(varnum) {}
+        std::unique_ptr<AstNode> operator()(std::unique_ptr<AstNode> node) const override {
+            if (AstVariable* var = dynamic_cast<AstVariable*>(node.get())) {
+                // Rename the variable
+                AstVariable* newVar = var->clone();
+                std::stringstream newName;
+                newName << "<inlined_" << var->getName() << "_" << varnum << ">";
+                newVar->setName(newName.str());
+                return std::unique_ptr<AstNode>(newVar);
+            }
+            node->apply(*this);
+            return node;
+        }
+    };
+
+    VariableRenamer update(inlineCount);
+    atomClause->apply(update);
+
+    inlineCount++;
+
+    // Get the constraints needed to unify the two atoms
+    NullableVector<std::pair<AstArgument*, AstArgument*>> res = unifyAtoms(atomClause->getHead(), atom);
+    if (res.isValid()) {
+        changed = true;
+        for (std::pair<AstArgument*, AstArgument*> pair : res.getVector()) {
+            constraints.push_back(new AstConstraint(BinaryConstraintOp::EQ,
```

Comment:
This is one of the cases where std::unique_ptr vs std::make_unique actually makes a difference to the code. The compiler can order these in a way that is not exception safe. Instead, use
```
new AstConstraint(BinaryConstraintOp::EQ,
std::make_unique<AstArgument>(pair.first->clone()),
std::make_unique<AstArgument>(pair.second->clone()))
```

Reply:
I'm not sure what to use make_unique for in this case (and the ones below), since the argument's just being cloned (and a new one isn't being made).